### PR TITLE
Mk UI

### PR DESCRIPTION
**Pull request by masv3971**

### CODE REVIEW

File: internal/ui/httpserver/service.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	config        *model.Cfg
+	logger        *logger.Log
+	tp            *trace.Tracer
+	server        *http.Server
+	apiv1         Apiv1
+	gin           *gin.Engine
+	sessionConfig *sessionConfig
+}
+
+type sessionConfig struct {
```
Comment (masv3971) [inline internal/ui/httpserver/service.go:32]: use // for comments, place comments before type name.

Change name of types in struct;
sessionConfig.sessionName is a bit repetitive.

File: internal/ui/httpserver/service.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+// New creates a new httpserver service
+func New(ctx context.Context, config *model.Cfg, api *apiv1.Client, tracer *trace.Tracer, logger *logger.Log) (*Service, error) {
+	s := &Service{
+		config: config,
+		logger: logger,
+		tp:     tracer,
+		apiv1:  api,
+		server: &http.Server{
+			Addr: config.UI.APIServer.Addr,
+			/* TODO: set all server configutations below and remove the same set(s) after s.server.Handler = s.gin ?
```
Comment (masv3971) [inline internal/ui/httpserver/service.go:53]: Remove comments.

File: internal/ui/httpserver/service.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	rgRoot := s.gin.Group("/")
+	s.regEndpoint(ctx, rgRoot, http.MethodPost, "/login", s.endpointLogin)
+	s.regEndpoint(ctx, rgRoot, http.MethodGet, "/health", s.endpointStatus)
+
+	rgSecure := rgRoot.Group("secure")
+	rgSecure.Use(s.authRequired)
+	s.regEndpoint(ctx, rgSecure, http.MethodDelete, "/logout", s.endpointLogout)
+	s.regEndpoint(ctx, rgSecure, http.MethodGet, "/user", s.endpointUser)
+
+	///apigw
```
Comment (masv3971) [inline internal/ui/httpserver/service.go:130]: remove one /

File: internal/ui/httpserver/service.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	s.regEndpoint(ctx, rgSecure, http.MethodGet, "/apigw/health", s.endpointAPIGWStatus)
```
Comment (masv3971) [inline internal/ui/httpserver/service.go:131]: Consider routergroup for apigw, mockas, etc..

File: internal/ui/httpserver/service.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	s.regEndpoint(ctx, rgSecure, http.MethodPost, "/apigw/portal", s.endpointPortal)
+
+	//mockas
+	s.regEndpoint(ctx, rgSecure, http.MethodPost, "/mockas/mock/next", s.endpointMockNext)
+
+	// Run http server
+	go func() {
+		//TODO: add tls support (see service apigw) + sessionSecure must be set to true
```
Comment (masv3971) [inline internal/ui/httpserver/service.go:139]: Fix comment, or remove it.

File: internal/ui/httpserver/middleware.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+func (s *Service) middlewareUserSession(ctx context.Context, cfg *model.Cfg) gin.HandlerFunc {
+	return setupSessionMiddleware(cfg, s)
+}
+
+func setupSessionMiddleware(cfg *model.Cfg, s *Service) gin.HandlerFunc {
+	store := configureSessionStore(cfg, s)
+	return sessions.Sessions(s.sessionConfig.sessionName, store)
+}
+
+func configureSessionStore(cfg *model.Cfg, s *Service) sessions.Store {
```
Comment (masv3971) [inline internal/ui/httpserver/middleware.go:70]: This does not look like a middleware, more like a init process for the store. Please explain.

File: cmd/ui/main.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	"time"
+	"vc/internal/ui/apiv1"
+	"vc/internal/ui/httpserver"
+	"vc/pkg/configuration"
+	"vc/pkg/logger"
+	"vc/pkg/trace"
+)
+
+func init() {
+	gob.Register(time.Time{})
```
Comment (masv3971) [inline cmd/ui/main.go:19]: Why do you need this? Can't see any use of it in this package.

File: internal/ui/apiv1/vc_base_client.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+		return nil, err
+	}
+
+	defer c.closeBody(resp)
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var jsonResp map[string]interface{}
```
Comment (masv3971) [inline internal/ui/apiv1/vc_base_client.go:69]: Same as line 98.

**Review by masv3971 — COMMENTED** [review]
In general, comments should be before method;

// Foo does that or that.
func (x *XX)Foo(){...}

// Bar does that...
type Bar struct {
   // Bis does this
   Bis string 
}

File: internal/ui/httpserver/middleware.go (hunk 6 above)
Comment (matskramer) [reply internal/ui/httpserver/middleware.go:70]: func Sessions(name string, store Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		s := &session{name, c.Request, store, nil, false, c.Writer}
		c.Set(DefaultKey, s)
		defer context.Clear(c.Request)
		c.Next()
	}
}

File: cmd/ui/main.go (hunk 7 above)
Comment (matskramer) [reply cmd/ui/main.go:19]: The code is an initialization function that registers the time.Time type with the gob package to enable the serialization and deserialization of time.Time values. 

Without I get an error: "2024-05-22T12:44:42.420Z	ERROR	vc_ui.httpserver	Failed to save session (and send cookie) during login	{"error": "securecookie: error - caused by: securecookie: error - caused by: gob: type not registered for interface: time.Time"}"

File: internal/ui/apiv1/handlers.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	"vc/pkg/model"
+)
+
+func (c *Client) Status(ctx context.Context, req *apiv1_status.StatusRequest) (*apiv1_status.StatusReply, error) {
+	probes := model.Probes{}
+	status := probes.Check("ui")
+	return status, nil
+}
+
+func (c *Client) Login(ctx context.Context, req *LoginRequest) (*LoggedinReply, error) {
```
[the remark this answers is not in the document]
Comment (matskramer) [reply internal/ui/apiv1/handlers.go:17]: Don't understand this comment?

File: cmd/ui/main.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	tracer, err := trace.New(ctx, cfg, log, "vc", "ui")
+	if err != nil {
+		panic(err)
+	}
+
+	apigwClient := apiv1.NewAPIGWClient(cfg, tracer, log.New("ui_apiwg_client"))
```
[the remark this answers is not in the document]
Comment (matskramer) [reply cmd/ui/main.go]: FIXED

File: cmd/ui/main.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+	mockasClient := apiv1.NewMockASClient(cfg, tracer, log.New("ui_mockas_client"))
```
[the remark this answers is not in the document]
Comment (matskramer) [reply cmd/ui/main.go]: FIXED

File: dockerfiles/ui_worker [hunk 12]
```diff
@@ -0,0 +1,38 @@
+# Compile
+FROM docker.sunet.se/dc4eu/gobuild:latest AS builder
+
+COPY . .
+ARG SERVICE_NAME
+
+#RUN make swagger
```
[the remark this answers is not in the document]
Comment (matskramer) [reply dockerfiles/ui_worker]: FIXED

File: dockerfiles/ui_worker [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+ARG SERVICE_NAME
+
+WORKDIR /
+
+RUN apt-get update && apt-get install -y curl procps iputils-ping less
+RUN rm -rf /var/lib/apt/lists/*
+
+COPY --from=builder /go/src/app/bin/vc_${SERVICE_NAME} /vc_service
+COPY --from=builder /go/src/app/internal/ui/static /static/
+#COPY --from=builder /go/src/app/docs /docs
```
[the remark this answers is not in the document]
Comment (matskramer) [reply dockerfiles/ui_worker]: FIXED

File: internal/ui/httpserver/service.go (hunk 4 above)
Comment (matskramer) [reply internal/ui/httpserver/service.go]: FIXED

File: internal/ui/apiv1/vc_base_client.go (hunk 8 above)
Comment (matskramer) [reply internal/ui/apiv1/vc_base_client.go]: FIXED

File: docker-compose.yaml [hunk 14]
```diff
@@ -18,6 +18,20 @@ services:
...
       - "VC_CONFIG_YAML=config.yaml"
 
+  ui:
+    container_name: "vc_dev_ui"
+    image: docker.sunet.se/dc4eu/ui:latest
+    restart: always
+    volumes:
+      - ./dev_config_docker.yaml:/config.yaml:ro
+    depends_on:
+      - apigw
```
[the remark this answers is not in the document]
Comment (matskramer) [reply docker-compose.yaml:28]: Decided not to add a required dependency to mockas

[the remark this answers is not in the document]
Comment (masv3971) [reply docker-compose.yaml:28]: UI will mainly be used in some demo setting, I suggest to add mockas as a depends_on.

File: internal/ui/httpserver/middleware.go (hunk 6 above)
Comment (masv3971) [reply internal/ui/httpserver/middleware.go:70]: I don't get it.
Is func Sessions() something you want to use or what is it?

File: internal/ui/apiv1/handlers.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (matskramer) [reply internal/ui/apiv1/handlers.go:26]: FIXED

File: internal/ui/apiv1/vc_base_client.go [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+	"vc/pkg/logger"
+	"vc/pkg/trace"
+)
+
+const (
```
[the remark this answers is not in the document]
Comment (matskramer) [reply internal/ui/apiv1/vc_base_client.go]: FIXED

File: internal/ui/apiv1/vc_base_client.go [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	defer c.closeBody(resp)
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var jsonResp map[string]interface{}
```
[the remark this answers is not in the document]
Comment (matskramer) [reply internal/ui/apiv1/vc_base_client.go]: FIXED

File: internal/ui/httpserver/service.go (hunk 3 above)
Comment (matskramer) [reply internal/ui/httpserver/service.go]: FIXED

File: internal/ui/apiv1/client.go [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+	"vc/pkg/model"
+	"vc/pkg/trace"
+)
+
+// Client holds the public api object
+type Client struct {
+	cfg     *model.Cfg
+	tp      *trace.Tracer
+	log     *logger.Log
+	apigwc  *APIGWClient
```
[the remark this answers is not in the document]
Comment (matskramer) [reply internal/ui/apiv1/client.go]: FIXED

File: internal/ui/httpserver/middleware.go (hunk 6 above)
Comment (matskramer) [reply internal/ui/httpserver/middleware.go:70]: It's the "glue" that makes the session and cookie handling work, see https://github.com/gin-contrib/sessions

File: docker-compose.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (matskramer) [reply docker-compose.yaml:28]: FIXED

File: internal/ui/httpserver/endpoints.go [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+	if err := session.Save(); err != nil { //This is also where the session cookie is created by gin
+		s.logger.Error(err, "Failed to save session (and send cookie) during login")
+		return nil, err
+	}
+
+	return reply, nil
+}
+
+func (s *Service) endpointLogout(ctx context.Context, c *gin.Context) (any, error) {
+	//gets the session based on session-ID in session cookie (handled by gin)
```
[the remark this answers is not in the document]
Comment (matskramer) [reply internal/ui/httpserver/endpoints.go]: FIXED

File: internal/ui/httpserver/middleware.go [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+	return setupSessionMiddleware(cfg, s)
+}
+
+func setupSessionMiddleware(cfg *model.Cfg, s *Service) gin.HandlerFunc {
+	store := configureSessionStore(cfg, s)
+	return sessions.Sessions(s.sessionConfig.name, store)
+}
+
+func configureSessionStore(cfg *model.Cfg, s *Service) sessions.Store {
+	//The first parameter is used to encrypt and decrypt cookies.
```
[the remark this answers is not in the document]
Comment (matskramer) [reply internal/ui/httpserver/middleware.go]: FIXED

File: internal/ui/httpserver/service.go [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+	apiv1         Apiv1
+	gin           *gin.Engine
+	sessionConfig *sessionConfig
+}
+
+// sessionConfig... values is also used for the session cookie
+type sessionConfig struct {
+	//if changed, the web (javascript) must also be updated with the new name
+	name string
+	// after this time with inactivity the session is auto removed from session storage - also the MaxAge value for the cookie
```
[the remark this answers is not in the document]
Comment (matskramer) [reply internal/ui/httpserver/service.go]: FIXED

File: internal/ui/httpserver/service.go [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+	inactivityTimeoutInSeconds int
+	path                       string
+	httpOnly                   bool
+	secure                     bool
+	sameSite                   http.SameSite
+	//key to retrive the username for the logged in user from the session (not stored in any cookie)
```
[the remark this answers is not in the document]
Comment (matskramer) [reply internal/ui/httpserver/service.go]: FIXED

File: pkg/model/config.go [hunk 22]
```diff
@@ -135,6 +135,23 @@ type OTEL struct {
 	Type string `yaml:"type" validate:"required"`
 }
 
+type UI struct {
+	APIServer APIServer `yaml:"api_server" validate:"required"`
+	// Login: user and pwd
```
[the remark this answers is not in the document]
Comment (matskramer) [reply pkg/model/config.go]: FIXED

File: internal/ui/apiv1/handlers.go [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+	return status, nil
+}
+
+type LoginRequest struct {
+	Username string `json:"username" binding:"required"`
+	Password string `json:"password" binding:"required"`
+}
+type LoggedinReply struct {
+	Username     string    `json:"username" binding:"required"`
+	LoggedInTime time.Time `json:"logged_in_time" binding:"required"` //time.Time encoded to JSON will use the RFC3339 format by default, which is essentially ISO 8601 (e.g., "2024-05-09T14:00:00Z"
```
[the remark this answers is not in the document]
Comment (matskramer) [reply internal/ui/apiv1/handlers.go]: FIXED

File: internal/ui/apiv1/vc_base_client.go [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+	"vc/pkg/logger"
+	"vc/pkg/trace"
+)
+
+// VCBaseClient Defines a base http(s) client for a service in the vc (verifiable credential) domain (for example: apigw, datastore, etc.)
```
[the remark this answers is not in the document]
Comment (matskramer) [reply internal/ui/apiv1/vc_base_client.go]: FIXED

File: internal/ui/httpserver/service.go [hunk 25]
```diff
@@ -113,16 +113,16 @@ func New(ctx context.Context, config *model.Cfg, api *apiv1.Client, tracer *trac
...
 	s.regEndpoint(ctx, rgRoot, http.MethodGet, "health", s.endpointStatus)
 
-	rgSecure := rgRoot.Group("secure")
+	rgSecure := rgRoot.Group("secure", s.middlewareAuthRequired(ctx))
 	rgSecure.Use(s.middlewareAuthRequired(ctx))
 	s.regEndpoint(ctx, rgSecure, http.MethodDelete, "logout", s.endpointLogout)
 	s.regEndpoint(ctx, rgSecure, http.MethodGet, "user", s.endpointUser)
 
-	rgAPIGW := rgSecure.Group("apigw")
+	rgAPIGW := rgSecure.Group("apigw", s.middlewareAuthRequired(ctx))
```
Comment (masv3971) [inline internal/ui/httpserver/service.go:121]: Doesn't middleware get inherent from previous router group?  Since rgAPIGW inherent from rgSecure.

Comment (matskramer) [reply internal/ui/httpserver/service.go:121]: FIXED

**Pull request merged**